intellij/src/saros/intellij/SarosComponent.java

```diff
@@ -3,27 +3,25 @@
 import com.intellij.openapi.actionSystem.KeyboardShortcut;
 import com.intellij.openapi.keymap.Keymap;
 import com.intellij.openapi.keymap.KeymapManager;
-import com.intellij.openapi.project.Project;
 import java.awt.event.KeyEvent;
 import java.io.InputStream;
 import javax.swing.KeyStroke;
 import org.apache.log4j.PropertyConfigurator;
 import org.apache.log4j.helpers.LogLog;
-import org.jetbrains.annotations.NotNull;
 
 /**
- * Component that is initalized when a project is loaded. It initializes the logging, shortcuts and
+ * Component that is initialized when a project is loaded. It initializes the logging, shortcuts and
  * the {@link IntellijProjectLifecycle} singleton.
  */
-public class SarosComponent implements com.intellij.openapi.components.ProjectComponent {
+public class SarosComponent {
```

Comment:
https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/openapi/components/ApplicationComponent.java

* Instead of {@link #initComponent()} please use {@link com.intellij.util.messages.MessageBus} and corresponding topics.
 * Instead of {@link #disposeComponent()} please use {@link com.intellij.openapi.Disposable}.

I would prefer to use such methods and not doing the heavy in the CTOR,

Looking at the code, is the session closed when you shutdown IntelliJ, i.e not stopping the session manually ?

Reply:
> I would prefer to use such methods and not doing the heavy in the CTOR,

I also considered the approach (as it is actively requested by the documentation), but couldn't find a good topic to use. The way I understand it, the thought behind moving this to message bus calls is reducing the load while initializing the application/project, i.e. only load the plugins when they are really needed (as stated in the beginning of the javadoc).

But I am not sure our plugin fits into this model as it is not bound to a specific functionality but rather offers a service to the user. I could, in theory, change it so that the plugin is only loaded when the user opens a project for the first time, but I am not sure this really improves the behavior of the plugin. (Furthermore, this restricts us from starting a session from the project menu. Even though this is currently not implemented, it could be a useful entry point when wanting to accept sessions on entire projects, especially when they don't exist locally yet.)

This would still leave us with the option of using `initComponent()`, but I am not sure whether this is an improvement.

> Looking at the code, is the session closed when you shutdown IntelliJ, i.e not stopping the session manually ?

No, it is not (currently).

As far as I know, terminating IntelliJ triggers project closed listeners, which we can use for this purpose.
This logic was initially part of #539, but I decided to move it to a separate PR. Since that decision, I have worked on it a bit, but still couldn't find a good solution avoiding the synchronous session end (session end should not be called on the EDT, but I need to ensure that no activities for the closed project are handled; implementing this with a filter in the core did not prove as effective as initially thought). I will open a WIP PR for this logic, showing off the approach and further explaining the remaining issues.

We also don't handle the disposed case explicitly. My assumption was that the plugin is torn down automatically when the application is terminated/torn down.

Reply:
@srossbach Do you have any issues with the current state of the PR (besides the  CTOR issue)? Otherwise, I would proceed with merging this.

If you would like, I can also open an issue for the CTOR issue so that we can keep it in mind (and have another look at it later on).

Reply:
Creating an issue stating that the session is not gracefully shutdown (it is not shutdown at all) when you close the IntelliJ IDE process should be ok.

Reply:
Open #582 for this.